flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessageClientDecoder.java

```diff
@@ -0,0 +1,142 @@
+ * +-----------------------------------+--------------------------------+
+ * | FRAME_HEADER ||  MESSAGE_HEADER   |     DATA BUFFER (Optional)     |
+ * +-----------------------------------+--------------------------------+
+ *
+ * This decoder decodes the frame header and delegates the other following work
+ * to corresponding message parsers according to the message type. During the process
+ * of decoding, the decoder and parsers try best to eliminate copying. For the frame
+ * header and message header, it only cumulates data when they span multiple input buffers.
+ * For the buffer part, it copies directly to the input channels to avoid future copying.
+ *
+ * The format of the frame header is
+ * +------------------+------------------+--------+
+ * | FRAME LENGTH (4) | MAGIC NUMBER (4) | ID (1) |
+ * +------------------+------------------+--------+
+ */
+public class NettyMessageClientDecoder extends ChannelInboundHandlerAdapter {
+
+	/** The message parser for buffer response. */
+    private final NettyMessageDecoderDelegate bufferResponseDecoderDelegate;
+
+    /** The message parser for other messages other than buffer response. */
+	private final NettyMessageDecoderDelegate nonBufferResponseDecoderDelegate;
+
+	/** The cumulation buffer for the frame header part. */
+	private ByteBuf frameHeaderBuffer;
+
+	/**
+	 * The chosen message parser for the current message. If it is null, then
+	 * we are decoding the frame header part, otherwise we are decoding the actual
+	 * message.
+	 */
+	private NettyMessageDecoderDelegate currentDecoderDelegate;
+
+    NettyMessageClientDecoder(NetworkClientHandler networkClientHandler) {
+        this.bufferResponseDecoderDelegate = new BufferResponseDecoderDelegate(networkClientHandler);
+        this.nonBufferResponseDecoderDelegate = new NonBufferResponseDecoderDelegate();
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+
+        bufferResponseDecoderDelegate.onChannelActive(ctx);
+        nonBufferResponseDecoderDelegate.onChannelActive(ctx);
+
+		frameHeaderBuffer = ctx.alloc().directBuffer(FRAME_HEADER_LENGTH);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        super.channelInactive(ctx);
+
+		bufferResponseDecoderDelegate.close();
+		nonBufferResponseDecoderDelegate.close();
+
+		frameHeaderBuffer.release();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        ByteBuf data = (ByteBuf) msg;
+
+        try {
+            while (data.isReadable()) {
+				if (currentDecoderDelegate == null) {
+					decodeFrameHeader(data);
+				}
+
+				if (data.isReadable() && currentDecoderDelegate != null) {
+					NettyMessageDecoderDelegate.ParseResult result = currentDecoderDelegate.onChannelRead(data);
+
+					if (result.finished) {
+						if (result.message != null) {
```

Comment:
Based on the [comment](https://github.com/apache/flink/pull/7368/files#r381770789) , we can refactor the `ParseResult` to make sure the message is always `nonnull` if finished.

Reply:
Removed the `nonnull` annotation.